core/dbt/task/manage.py

```diff
@@ -0,0 +1,77 @@
+# coding=utf-8
+from typing import Dict, Set, Tuple
+
+from .compile import CompileTask
+from .runnable import ManifestTask
+from dbt.exceptions import warn_or_error, ValidationException
+from dbt.adapters.factory import get_adapter
+from dbt.contracts.graph.parsed import (
+    ParsedModelNode,
+)
+from dbt.contracts.project import PruneModelsAction
+
+
+class ManageTask(CompileTask):
+    def run(self):
+        ManifestTask._runtime_initialize(self)
+        models_in_codebase = self.manifest.nodes.keys()
+        adapter = get_adapter(self.config)
+
+        with adapter.connection_named("master"):
+            required_schemas = self.get_model_schemas(adapter, models_in_codebase)
+            self.populate_adapter_cache(adapter, required_schemas)
+
+            adapter.clear_transaction()
+            self._prune_models(adapter)
+
+    def _prune_models(self, adapter):
+        self._assert_schema_uniqueness()
+
+        if len(self.config.managed_schemas) == 0:
+            warn_or_error("No schema's configured to manage")
+            return
+
+        models_in_codebase: Set[Tuple[str, str, str]] = set(
+            (n.database, n.schema, n.name)
+            for n in self.manifest.nodes.values()
+            if isinstance(n, ParsedModelNode)
+        )
+
+        for config in self.config.managed_schemas:
+            generated_schema_name = adapter.execute_macro(
+                "generate_schema_name", kwargs={"custom_schema_name": config.schema}
+            )
+
+            models_in_database: Dict[Tuple[str, str, str], str] = {
+                (config.database, generated_schema_name, relation.identifier): relation
+                for relation in adapter.list_relations(config.database, generated_schema_name)
+            }
+            if len(models_in_database) == 0:
+                warn_or_error(
+                    f"No modules in managed schema '{generated_schema_name}' for database '{config.database}'"
+                )
+
+            should_act_upon = models_in_database.keys() - models_in_codebase
+
+            for (target_database, target_schema, target_identifier) in should_act_upon:
+                target_action = config.prune_models or PruneModelsAction.SKIP
+                if target_action == PruneModelsAction.WARN:
+                    warn_or_error(
+                        f"Found unused model {target_database}.{target_schema}.{target_identifier}"
+                    )
+                elif target_action == PruneModelsAction.DROP:
+                    adapter.drop_relation(
+                        models_in_database[(target_database, target_schema, target_identifier)]
+                    )
```

Comment:
This will run the queries in sequence, instead of in parallel threads (up to the number of configured `--threads`). That's probably okay for a first cut, but it would be nice to make use of parallelization, since that's also a big benefit of defining this as a built-in task (written in Python) instead of a `run-operation` Jinja macro.

There are two approaches I could see to make that happen:
1. Use the multi-threaded `executor` utility + `concurrent.futures`, same as we do for (e.g.) [caching](https://github.com/dbt-labs/dbt-core/blob/5e4e917de5f4da7b24862a2d9b14a2709686e3ff/core/dbt/adapters/base/impl.py#L393-L408) + [cataloging](https://github.com/dbt-labs/dbt-core/blob/5e4e917de5f4da7b24862a2d9b14a2709686e3ff/core/dbt/adapters/base/impl.py#L1057-L1069) queries.
2. Turn this task into a real "graph runnable" task. Create a graph queue from the `should_act_upon` set, define a "runner" type, and define an `execute` method on that "runner." This follows a better-established pattern. The tricky part is, all other tasks expect to construct that graph queue from the dbt DAG; the objects in `should_act_upon` are (by definition) _not_ members of the dbt DAG.

Reply:
How important is this? I understand that it might increase performance, and that its more in line with `—threads`. On the other hand, it adds complexity. Also, the database might be the bottleneck anyhow, since they have to do a lot of integrity checks and actual data dropping. It is not clear to me whether dropping many tables simultaneously would be quicker in the database. 

I’d propose to leave this out in a first version to keep it small and simple.

Reply:
I would say to leave out the parallel processing for now, given that the config is going to go through a refactoring, this feature still needs to be in the hands of users to get feedback on usage and the drop actions are probably not a bottleneck when it comes to performance.

Resolving this as 'nice to have'